### PR TITLE
GH-96764: rewrite `asyncio.wait_for` to use `asyncio.timeout`

### LINKED ISSUE #81839: In some cases asyncio.wait_for can lead to socket leak.
BPO | [37658](https://bugs.python.org/issue37658)
--- | :---
Nosy | @ods, @asvetlov, @cjerdonek, @elprans, @ambv, @1st1, @miss-islington, @Dreamsorcerer, @aaliddell, @kumaraditya303
PRs | <li>python/cpython#21894</li><li>python/cpython#21964</li><li>python/cpython#21965</li><li>python/cpython#21969</li><li>python/cpython#26097</li><li>python/cpython#29202</li><li>python/cpython#29832</li><li>python/cpython#29831</li>
Files | <li>[example.py](https://bugs.python.org/file48500/example.py "Uploaded as text/plain at 2019-07-23.12:16:16 by Nikita Ilyasov"): script that can reproduce the problem</li><li>[ssl_circular_refs.png](https://bugs.python.org/file50061/ssl_circular_refs.png "Uploaded as image/png at 2021-05-24.13:02:10 by @ods")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-07-23.12:16:17.363>
labels = ['expert-asyncio', '3.7', 'performance']
title = 'In some cases asyncio.wait_for can lead to socket leak.'
updated_at = <Date 2022-03-11.17:35:41.865>
user = 'https://bugs.python.org/NikitaIlyasov'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-11.17:35:41.865>
actor = 'dreamsorcerer'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['asyncio']
creation = <Date 2019-07-23.12:16:17.363>
creator = 'Nikita Ilyasov'
dependencies = []
files = ['48500', '50061']
hgrepos = []
issue_num = 37658
keywords = ['patch']
message_count = 18.0
messages = ['348328', '348329', '375937', '375942', '375947', '379542', '393566', '393567', '393584', '393586', '393687', '394249', '404998', '407248', '407252', '407373', '414797', '414925']
nosy_count = 12.0
nosy_names = ['ods', 'asvetlov', 'chris.jerdonek', 'Elvis.Pranskevichus', 'lukasz.langa', 'yselivanov', 'miss-islington', 'Nikita Ilyasov', 'dreamsorcerer', 'aaliddell', 'nmatravolgyi', 'kumaraditya']
pr_nums = ['21894', '21964', '21965', '21969', '26097', '29202', '29832', '29831']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'resource usage'
url = 'https://bugs.python.org/issue37658'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

### LINKED ISSUE #86296: AsyncIO's wait_for can hide cancellation in a rare race condition
BPO | [42130](https://bugs.python.org/issue42130)
--- | :---
Nosy | @ods, @asvetlov, @cjerdonek, @1st1, @tvoinarovskyi, @Dreamsorcerer, @aaliddell, @crowbarz, @wjcferguson
PRs | <li>python/cpython#26097</li><li>python/cpython#28149</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-10-23.16:10:28.990>
labels = ['type-bug', '3.8', '3.9', '3.10', 'expert-asyncio']
title = "AsyncIO's wait_for can hide cancellation in a rare race condition"
updated_at = <Date 2021-12-02.09:55:13.171>
user = 'https://github.com/tvoinarovskyi'
```

bugs.python.org fields:
```python
activity = <Date 2021-12-02.09:55:13.171>
actor = 'dreamsorcerer'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['asyncio']
creation = <Date 2020-10-23.16:10:28.990>
creator = 'tvoinarovskyi'
dependencies = []
files = []
hgrepos = []
issue_num = 42130
keywords = ['patch']
message_count = 6.0
messages = ['379454', '379541', '379577', '393336', '407489', '407521']
nosy_count = 10.0
nosy_names = ['ods', 'asvetlov', 'chris.jerdonek', 'yselivanov', 'tvoinarovskyi', 'dreamsorcerer', 'aaliddell', 'nmatravolgyi', 'crowbarz', 'jamesf']
pr_nums = ['26097', '28149']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue42130'
versions = ['Python 3.8', 'Python 3.9', 'Python 3.10']
```

</p></details>

### Linked PRs
* gh-98607

### LINKED ISSUE #96764: Rewrite `asyncio.wait_for` using `asyncio.timeout`
Over the years a lot of issues have accumulated for `asyncio.wait_for` and the code has become complicated. The `asyncio.timeout` can be used to simplify this a lot and will also fix bugs which are already fixed or don't exist in `asyncio.timeout`. This rewrite won't be backported. `asyncio.wait_for` should be nothing more than a wrapper around `asyncio.timeout`.

`asyncio.wait_for`  issues: https://github.com/python/cpython/issues?q=is%3Aissue+is%3Aopen+wait_for+label%3Aexpert-asyncio

### Linked PRs
* gh-98518

### PR DESCRIPTION
**Pull request by kumaraditya303**
This PR changes `asyncio.wait_for` to use `asyncio.timeout` as its underlying implementation. It simplifies the code and makes it easy to understand the cancellation semantics as both `asyncio.timeout` and `asyncio.wait_for` behaves similarly. 

Fixes https://github.com/python/cpython/issues/86296
Fixes https://github.com/python/cpython/issues/81839
Fixes #96764

* Issue: gh-96764

### CODE REVIEW

**Comment by kumaraditya303** [comment]
This is more of a POC at this point. I think it is worth changing `wait_for` to use `timeout` in 3.12+. This seems to fix some of the open bugs of `wait_for`, I'll add those tests to make sure. 

cc @gvanrossum

File: Lib/asyncio/tasks.py [hunk 1]
```diff
@@ -439,63 +440,23 @@ async def wait_for(fut, timeout):
...
-
     if timeout is None:
         return await fut
 
     if timeout <= 0:
-        fut = ensure_future(fut, loop=loop)
+        fut = ensure_future(fut)
 
         if fut.done():
             return fut.result()
```
Comment (gvanrossum) [inline Lib/asyncio/tasks.py:469]: It works even without these two lines. :-)

However, this whole block (L443-L456) is somehow required, otherwise several things go wrong. I wonder if that points to a flaw in `timeout()`?

File: Lib/test/test_asyncio/test_futures2.py [hunk 2]
```diff
@@ -76,21 +76,6 @@ class CFutureTests(FutureTests, unittest.IsolatedAsyncioTestCase):
...
-        # in base_futures:_future_repr_info is absent
-        # See Also: https://bugs.python.org/issue42183
-
-        async def func():
-            return asyncio.all_tasks()
-
-        # The repr() call should not raise RecursiveError at first.
-        # The check for returned string is not very reliable but
-        # exact comparison for the whole string is even weaker.
-        self.assertIn('...', repr(await asyncio.wait_for(func(), timeout=10)))
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_futures2.py:92]: Why did you have to remove this test?

File: Lib/test/test_asyncio/test_waitfor.py [hunk 3]
```diff
@@ -256,7 +256,6 @@ async def inner():
         # Test that wait_for itself is properly cancellable
         # even when the initial task holds up the initial cancellation.
         task = asyncio.create_task(wait)
-        await asyncio.sleep(0.2)
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_waitfor.py:259]: Curious why the test fails with this sleep.

**Review by gvanrossum — COMMENTED** [review]
You got a bit farther than I did when I tried this, but I'm still concerned why you had to modify several tests. (Also, there's one test that fails in CI, but doesn't appear to fail when I run it locally.)

File: Lib/asyncio/tasks.py [hunk 4]
```diff
@@ -439,63 +440,23 @@ async def wait_for(fut, timeout):
...
-            # In case task cancellation failed with some
-            # exception, we should re-raise it
-            # See https://bugs.python.org/issue40607
-            try:
-                return fut.result()
-            except exceptions.CancelledError as exc:
-                raise exceptions.TimeoutError() from exc
-    finally:
-        timeout_handle.cancel()
+            raise TimeoutError from exc
```
Comment (gvanrossum) [inline Lib/asyncio/tasks.py:475]: Somewhere here (or in `_cancel_and_wait()`) we ought to call `uncancel()`, but only if _we_ are the party that called `cancel()`, and then only if an `uncancel` method actually exists. (It may not if it's a future, or if it's a task created by a 3rd party event loop that doesn't yet support `ucancel()`.)

File: Lib/test/test_asyncio/test_futures2.py (hunk 2 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py:92]: Updated, this test now only needs to test that no `RecursionError` is raised as now `asyncio.wait_for` no longer create new task.

File: Lib/asyncio/tasks.py (hunk 1 above)
Comment (kumaraditya303) [reply Lib/asyncio/tasks.py:469]: Not a flaw but a different design choice. I had pointed this out in https://github.com/python/cpython/pull/31394#discussion_r817952872 on the PR adding `timeout`.

TL;DR
- `asyncio.wait_for` raises `TimeoutError` if `<=0` is passed as deadline.
- `asyncio.timeout` disables deadline if `None` is passed.

Anyways, changing this is out of scope for this PR so marking resolved.

File: Lib/asyncio/tasks.py (hunk 4 above)
Comment (kumaraditya303) [reply Lib/asyncio/tasks.py:475]: Why is that required? I haven't used `uncancel` much so might be missing the context.

File: Lib/asyncio/tasks.py (hunk 1 above)
Comment (gvanrossum) [reply Lib/asyncio/tasks.py:469]: Correction, wait_for() also disables deadline if None is passed as timeout. It's the first two lines of code in this function. So as far as these two bullets are concerned they are actually equivalent.

So I think there's something else that this block (L443-L456) ensures that timeout() doesn't.

Comment (kumaraditya303) [reply Lib/asyncio/tasks.py:469]: Right, the only special case is `<=0` as deadline in `wait_for` which behaves differently. `test_wait_for_timeout_less_then_0_or_0_coroutine_do_not_started` fails if this special case is removed.

Comment (kumaraditya303) [reply Lib/asyncio/tasks.py:469]: Removed `None` special case.

File: Lib/asyncio/tasks.py (hunk 4 above)
Comment (gvanrossum) [reply Lib/asyncio/tasks.py:475]: Uncancel is the great mystery of asyncio. :-)

Seriously, when things like wait_for() or timeout() are nested, it is possible that two timeouts occur "simultaneously", and a task gets cancelled twice before its coroutine gets the CancelledError thrown in. It's no problem for the task (it should just clean up and reraise), but the inner timeout/wait_for call should _not_ raise a TimeoutError, but instead bubble up the CancelledError.

Why not raise TimeoutError? Because the outer thing might not be a timeout -- it might be a TaskGroup or some other thing. For example, we could have something like "happy eyeballs" that wants to cancel all other attempts once the first attempt has a result. But it might also set timeouts on other attempts. The cancellation should preempt the timeout.

To implement such cooperative cancellation schemes, Task.cancel() increments a counter, and code that cancels a task and later handles the cancellation is supposed to call uncancel(), which decrements the counter and returns the remaining value. If that's nonzero, there was another cancel that had priority (because its handling will happen in an outer construct).

The timeout() context manager (and TaskGroup() too) implements this collaborative scheme, I think wait_for() should too.

File: Lib/asyncio/tasks.py (hunk 1 above)
Comment (gvanrossum) [reply Lib/asyncio/tasks.py:469]: > Right, the only special case is `<=0` as deadline in `wait_for` which behaves differently. `test_wait_for_timeout_less_then_0_or_0_coroutine_do_not_started` fails if this special case is removed.

Ah, thanks for explaining. Using `wait_for(t, 0)`, `t` is cancelled before it is started, whereas when using `async with timeout(0)` a task is allowed to run until its first `await` or exit.

Now I'm wondering which design is better, but we should first make `wait_for()` work as it always did.

File: Misc/NEWS.d/next/Library/2022-10-22-09-26-43.gh-issue-96764.Dh9Y5L.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+:func:`asyncio.wait_for` now uses :func:`asyncio.timeout` as it's underlying implementation. Patch by Kumar Aditya.
```
Comment (gvanrossum) [inline Misc/NEWS.d/next/Library/2022-10-22-09-26-43.gh-issue-96764.Dh9Y5L.rst]:
```suggestion
:func:`asyncio.wait_for` now uses :func:`asyncio.timeout` as its underlying implementation. Patch by Kumar Aditya.
```

File: Lib/test/test_asyncio/test_waitfor.py [hunk 6]
```diff
@@ -289,6 +288,42 @@ async def test_cancel_blocking_wait_for(self):
...
+        # See https://github.com/python/cpython/pull/28149
+
+        async def return_42():
+            try:
+                await asyncio.sleep(10)
+            except asyncio.CancelledError:
+                return 42
+
+        res = await asyncio.wait_for(return_42(), timeout=0.1)
+        self.assertEqual(res, 42)
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_waitfor.py:278]: I still feel this should raise TimeoutError. We should probably discuss this back on GH-86296.

Comment (gvanrossum) [reply Lib/test/test_asyncio/test_waitfor.py:278]: Sorry, I take it back. It seems wait_for() currently works this way (at least since 3.9) and so does timeout().

File: Lib/test/test_asyncio/test_waitfor.py [hunk 7]
```diff
@@ -289,6 +288,42 @@ async def test_cancel_blocking_wait_for(self):
...
+
+    async def test_wait_for_issue86296(self):
+        # See https://github.com/python/cpython/issues/86296
+
+        async def inner():
+            return
+
+        reached_end = False
+
+        async def with_for_coro():
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_waitfor.py]: Maybe you meant this to be called `wait_for_coro`?

**Review by gvanrossum — APPROVED** [review]
LG (one naming nit).

File: Lib/test/test_asyncio/test_waitfor.py (hunk 7 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py]: fixed

File: Lib/test/test_asyncio/test_waitfor.py (hunk 3 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py:259]: It is because `wait_for` no longer creates a new task for `coroutine` and `asyncio.timeout` executes the return 42 always which is the expected behavior.

File: Lib/asyncio/tasks.py (hunk 4 above)
Comment (kumaraditya303) [reply Lib/asyncio/tasks.py:475]: Thanks for the explanation though I feel it would be better if you do this as you understand `uncancel`  better than me.
This PR only targets 3.12 so we have time to do this, so leaving this for another day (for now).

**Comment by gvanrossum** [comment]
I had hoped to first merge one of the alternatives that can be backported to 3.10, so we can declare this fixed in 3.10. Then we would merge this one on top of that, but only in 3.11 and main.

But we could also just do a custom fix for 3.10 based on one of @twisteroidambassador's PRs. (I wish you had linked this PR to the same issue rather than creating a new issue, since it's all related.)

**Comment by gvanrossum** [comment]
In particular, @twisteroidambassador has this PR: https://github.com/python/cpython/pull/98607

File: Lib/test/test_asyncio/test_waitfor.py (hunk 3 above)
Comment (gvanrossum) [reply Lib/test/test_asyncio/test_waitfor.py:259]: Oh so it will still work with the sleep but the sleep is no longer necessary to make the test pass? Good call then!

Comment (twisteroidambassador) [reply Lib/test/test_asyncio/test_waitfor.py:259]: This test fails if the sleep is not removed.

When the sleep is removed, `task` was created and then cancelled immediately, meaning it did not get a chance to run, thus the `wait_for()` call inside it was never executed. I believe this change invalidates the test.

Comment (gvanrossum) [reply Lib/test/test_asyncio/test_waitfor.py:259]: It's too late here to investigate more but I take it that you believe the test is now passing for the wrong reason? @kumaraditya303 There's something to that...

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py:259]: Perhaps I didn't make myself clear so here's a longer explanation.

In previous implementation of `wait_for`, it worked by creating a new future and then awaiting on it so it was possible to cancel `wait_for` and would cancel that future but in the new implementation with `timeout` it no longer creates a new task so cancelling is has the same effect as cancelling the coroutine `wait_for` is waiting for but since here the coroutine is suppressing the cancellation so it will not raise `CancelledError`  with `sleep` but return 42. I can remove the test if you want though but went I went with making smallest possible change to tests to make it pass.

Comment (twisteroidambassador) [reply Lib/test/test_asyncio/test_waitfor.py:259]: Yes, I believe the change made to this test means it's not actually testing `wait_for()` anymore, and it passes for the wrong reason.

If I understand correctly, this test requires that when `wait_for(inner_coro(), timeout)` *EDIT:* ~~times out~~ is cancelled, even if `inner_coro` ignores cancellation (by catching and not reraising CancelledError), `wait_for` still raises CancelledError. This PR modifies this behavior, and returns `inner_coro`'s result instead.

**Comment by kumaraditya303** [comment]
@gvanrossum Do you want to merge the this ones first? I would say we merge this first and then create a fix especially for 3.10 and 3.11 branches since main will be rewritten anyways by this. I would also like to review them individually.  

> (I wish you had linked this PR to the same issue rather than creating a new issue, since it's all related.)

We can say that my issue superseded all the other ones :)

**Comment by kumaraditya303** [comment]
Merging this first also has the benefit that we have something working without bugs to compare against but it's ok if you want to go the other way around.

File: Lib/test/test_asyncio/test_waitfor.py (hunk 3 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py:259]: Okay I'll remove the test. 

> This PR modifies this behavior, and returns inner_coro's result instead.

Correct!, since no future is created, `wait_for` does what the coroutine does.

Comment (twisteroidambassador) [reply Lib/test/test_asyncio/test_waitfor.py:259]: I think we want to preserve existing behavior, especially if there's a test specifically testing for it? @gvanrossum should decide, I believe.

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py:259]: The behavior does not makes sense, if you `wait_for` a coroutine and then cancel it and if the coroutine suppresses the cancellation and return something then `wait_for` should return it rather than throwing `CancelledError`. This is one of the bugs we are trying to fix here. It also aligns it with how `timeout` works.

See https://github.com/python/cpython/blob/5fece55bb90b3d6e46b5d27d60bb31cbf5e5b9fc/Lib/test/test_asyncio/test_waitfor.py#L265-L270

File: Lib/test/test_asyncio/test_futures2.py [hunk 8]
```diff
@@ -89,7 +89,7 @@ async def func():
         # The repr() call should not raise RecursiveError at first.
         # The check for returned string is not very reliable but
```
Comment (akulakov) [inline Lib/test/test_asyncio/test_futures2.py]: The 2-line comment can be removed.

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py]: fixed

File: Lib/test/test_asyncio/test_waitfor.py (hunk 3 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py:259]: @gvanrossum ^^

**Comment by kumaraditya303** [comment]
Revisiting this, tests for the bug reports linked are added and existing tests pass without any change except one. 
The only remaining thing is the removed `test_wait_for_self_cancellation` test which tests that `wait_for` creates a future and awaits on it and its cancellation.  See https://github.com/python/cpython/pull/98518#discussion_r1005269554

File: Lib/test/test_asyncio/test_futures2.py (hunk 2 above)
Comment (gvanrossum) [reply Lib/test/test_asyncio/test_futures2.py:92]: I think this defeats the purpose of the test. The test is about checking that *if* repr() of a Task ends up recursing into itself, the recursion is stopped (by reprlib inserting `...`). The test should probably be rewritten by creating an explicit task.
```
        self.assertIn('...', repr(await asyncio.wait_for(asyncio.Task(func()), timeout=10)))
```

File: Lib/test/test_asyncio/test_waitfor.py [hunk 9]
```diff
@@ -289,6 +262,42 @@ async def test_cancel_blocking_wait_for(self):
     async def test_cancel_wait_for(self):
         await self._test_cancel_wait_for(60.0)
 
+    async def test_wait_for_cancel_suppressed(self):
+        # See https://github.com/python/cpython/issues/84849
+        # See https://github.com/python/cpython/pull/28149
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_waitfor.py]: Are these the best links to explain what this tests for? The former is closed as duplicate of gh-86296, which is a very long and rambling discussion. The second is an open PR. Maybe instead of leading readers down those confusing paths, just write up in one sentence why we want the behavior tested for here? (Maybe also mention that in general we recommend against suppressing CancelledError, but *if* you do, this is the desired behavior.)

File: Lib/test/test_asyncio/test_waitfor.py [hunk 10]
```diff
@@ -289,6 +262,42 @@ async def test_cancel_blocking_wait_for(self):
...
+                await asyncio.sleep(10)
+            except asyncio.CancelledError:
+                return 42
+
+        res = await asyncio.wait_for(return_42(), timeout=0.1)
+        self.assertEqual(res, 42)
+
+
+    async def test_wait_for_issue86296(self):
+        # See https://github.com/python/cpython/issues/86296
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_waitfor.py]: Again, maybe explain the behavior we test for here?

Is it just that after sleep(0) we end up in at sleep(1) and not stuck in wait_for(inner(), ...)?

And if so, is this just a roundabout way to test that wait_for(inner(), ...) doesn't create a Task and doesn't actually make a roundtrip through the event loop since inner() returns immediately?

File: Lib/asyncio/tasks.py [hunk 11]
```diff
@@ -439,63 +440,21 @@ async def wait_for(fut, timeout):
 
     This function is a coroutine.
     """
```
Comment (gvanrossum) [inline Lib/asyncio/tasks.py]: Maybe add to the docstring that if the task catches the cancellation and returns a value instead, that value is returned.

And the mention of `shield` reminds me that there's no test that demonstrate `wait_for(shield(coro()), ...)`. So maybe add a few? (One for each path through the code; I count about 4 paths.)

File: Lib/asyncio/tasks.py [hunk 12]
```diff
@@ -439,63 +440,21 @@ async def wait_for(fut, timeout):
...
-    loop = events.get_running_loop()
-
-    if timeout is None:
-        return await fut
 
-    if timeout <= 0:
-        fut = ensure_future(fut, loop=loop)
+    if timeout is not None and timeout <= 0:
```
Comment (gvanrossum) [inline Lib/asyncio/tasks.py:465]: Please add a comment explaining the subtle difference between `wait_for` and `timeout` when `timeout <= 0` (see the resolved thread below) so that future maintainers won't try to "optimize" it and then spend time understanding why the test fails?

**Review by gvanrossum — COMMENTED** [review]
Thanks for your patience! I have a few small suggestions and comments.

File: Lib/asyncio/tasks.py (hunk 11 above)
Comment (kumaraditya303) [reply Lib/asyncio/tasks.py]: Added `shield` tests.

File: Lib/test/test_asyncio/test_waitfor.py (hunk 9 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py]: done

File: Lib/asyncio/tasks.py (hunk 12 above)
Comment (kumaraditya303) [reply Lib/asyncio/tasks.py:465]: >  that future maintainers won't try to "optimize" it and then spend time understanding why the test fails?

Added comment but I am not leaving anytime soon ;)

File: Lib/test/test_asyncio/test_waitfor.py (hunk 10 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py]: Added comments, the point is that the task should get cancelled and the `asyncio.sleep(1)` is not run.

> And if so, is this just a roundabout way to test that wait_for(inner(), ...) doesn't create a Task and doesn't actually make a roundtrip through the event loop since inner() returns immediately?

Yes

File: Lib/test/test_asyncio/test_waitfor.py [hunk 13]
```diff
@@ -289,6 +262,94 @@ async def test_cancel_blocking_wait_for(self):
...
+                return 42
+
+        res = await asyncio.wait_for(return_42(), timeout=0.1)
+        self.assertEqual(res, 42)
+
+
+    async def test_wait_for_issue86296(self):
+        # GH-86296: The task should get cancelled and not run to completion.
+        # inner completes in one cycle of the event loop so it
+        # complete before the task is cancelled.
```
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_waitfor.py]:
```suggestion
        # completes before the task is cancelled.
```

File: Lib/asyncio/tasks.py [hunk 14]
```diff
@@ -438,9 +438,16 @@ async def wait_for(fut, timeout):
...
 
+    If the task supresses the cancellation and returns a value instead,
+    that value is returned.
+
     This function is a coroutine.
     """
 
+    # When timeout <= 0, `asyncio.timeout` raises `TimeoutError`
+    # whereas `wait_for` disables the timeout so we special case
+    # it here to preserve the behavior.
```
Comment (gvanrossum) [inline Lib/asyncio/tasks.py]: The way I read the code following, it is not "disabling" the timeout, rather it is cancelling the future (`_cancel_and_wait`). I know there is a difference between this and what timeout() does, but I still don't understand the difference. :-( Can you explain it to me?

File: Lib/test/test_asyncio/test_waitfor.py [hunk 15]
```diff
@@ -277,27 +279,77 @@ async def return_42():
...
+        async def coro():
+            await asyncio.sleep(0.1)
+            return 'done'
+
+        task = asyncio.create_task(coro())
+        with self.assertRaises(asyncio.TimeoutError):
+            await asyncio.wait_for(asyncio.shield(task), timeout=0)
+
+        self.assertFalse(task.done())
+        self.assertFalse(task.cancelled())
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_waitfor.py:330]: This seems to contradict the comment -- if it's not done, it will eventually run, right? So shield() is cancelled but the task is allowed to run in the background.

Is this the same for using timeout() directly?

File: Lib/test/test_asyncio/test_waitfor.py [hunk 16]
```diff
@@ -277,27 +279,77 @@ async def return_42():
...
+            await asyncio.sleep(1)
+            return 'done'
+
+        task = asyncio.create_task(coro())
+        with self.assertRaises(asyncio.TimeoutError):
+            await asyncio.wait_for(asyncio.shield(task), timeout=0.1)
+
+        self.assertFalse(task.done())
+        self.assertFalse(task.cancelled())
+        self.assertEqual(await task, "done")
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_waitfor.py:363]: This makes the test very slow, it will wait 1 second.

**Review by gvanrossum — COMMENTED** [review]
Better, but I'm still unsure about the difference between wait_for() and timeout() if timeout<=0.

File: Lib/test/test_asyncio/test_waitfor.py (hunk 16 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py:363]: Reduced to 0.1.

File: Lib/test/test_asyncio/test_waitfor.py (hunk 15 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py:330]: I have elaborated it more, should be easier to understand now. The task is scheduled to run in background, the updated test verifies it.

File: Lib/asyncio/tasks.py (hunk 14 above)
Comment (kumaraditya303) [reply Lib/asyncio/tasks.py]: That comment is wrong, sorry for that, removed now.

This is the simplest case I can think of which explains the behaviour:

```py
import asyncio


async def test_timeout():
    func_started = False

    async def func():
        nonlocal func_started
        func_started = True

    async with asyncio.timeout(0):
        await func()

    assert func_started

asyncio.run(test_timeout())

async def test_waitfor():
    func_started = False

    async def func():
        nonlocal func_started
        func_started = True

    try:
        await asyncio.wait_for(func(), 0)
    except asyncio.TimeoutError:
        assert not func_started
    else:
        assert False


asyncio.run(test_waitfor())
```

On main both asserts pass.

On this branch if I remove the `timeout <= 0` branch then the assert in `test_waitfor` fails.

Comment (kumaraditya303) [reply Lib/asyncio/tasks.py]: I have added `test_waitfor` code as comment as why the special case is needed.

File: Lib/asyncio/tasks.py [hunk 17]
```diff
@@ -437,65 +438,44 @@ async def wait_for(fut, timeout):
 
     If the wait is cancelled, the task is also cancelled.
 
+    If the task supresses the cancellation and returns a value instead,
+    that value is returned.
+
     This function is a coroutine.
     """
-    loop = events.get_running_loop()
+    # The special case for timeout <= 0 is for the follwing case:
```
Comment (kumaraditya303) [inline Lib/asyncio/tasks.py]:
```suggestion
    # The special case for timeout <= 0 is for the following case:
```

File: Lib/test/test_asyncio/test_waitfor.py [hunk 18]
```diff
@@ -310,18 +310,30 @@ async def wait_for_coro():
...
+            shielded_task = asyncio.shield(task)
+            await asyncio.wait_for(shielded_task, timeout=0)
 
+        # Task is running in background
         self.assertFalse(task.done())
         self.assertFalse(task.cancelled())
+        self.assertTrue(shielded_task.cancelled())
+
+        # Wait for the task to complete
+        await asyncio.sleep(0.1)
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_waitfor.py:334]: Why not `await task`?

**Review by gvanrossum — COMMENTED** [review]
LG, just one nit.

File: Lib/test/test_asyncio/test_waitfor.py (hunk 18 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_waitfor.py:334]: That would start to run the task if it is not already scheduled, this verifies that the task was running in "background".

**Comment by kumaraditya303** [comment]
@gvanrossum Thanks for the review, this was a tough nut to crack.

File: Lib/test/test_asyncio/test_waitfor.py (hunk 18 above)
Comment (gvanrossum) [reply Lib/test/test_asyncio/test_waitfor.py:334]: I don’t believe there is a difference. Anything that cedes to the event loop will start tasks that are ready to run. And it is not the job of this test to verify that.
